pyro/ops/welford.py

```diff
@@ -0,0 +1,70 @@
+from collections import OrderedDict
+
+import torch
+
+
+class WelfordCovariance(object):
+    """
+    Implements Welford's online scheme for estimating (co)variance (see :math:`[1]`).
+    Useful for adapting diagonal and dense mass structures for HMC.
+
+    **References**
+
+    [1] `The Art of Computer Programming`,
+    Donald E. Knuth
+    """
+    def __init__(self, diagonal=True):
+        self.diagonal = diagonal
+        self.reset()
+
+    def _deltas(self, z):
+        deltas = OrderedDict()
+        for site_name, site_value in z.items():
+            mean = self.means.get(site_name, 0.)
+            delta_pre = site_value - mean
+            updated_mean = mean + delta_pre / self.n_samples
+            delta_post = site_value - updated_mean
+            deltas[site_name] = (updated_mean, delta_pre, delta_post)
+        return deltas
+
+    def _unroll(self, z):
+        unrolled = OrderedDict()
+        for site_name, rolled in z.items():
+            for i, x in enumerate(rolled.view(-1)):
```

Comment:
Use `.reshape(-1)` rather than `.view(-1)` to avoid issues with non-contiguous tensors.